Re-connecting WPCOM Account - Add the re-connecting flow

eason9487 opened the pull request:
### Changes proposed in this Pull Request:

It's a part of frontend in #775

- Add the `next` parameter to GET `jetpack/connect` API for indicating the redirect URI when back from Jetpack authorization.
- Make \<ConnectWPComAccountCard\> call GET `jetpack/connect` API with the `next` parameter.
- Add a new component \<ReconnectWPComAccount\> for re-connecting WPCOM account, and add the re-connecting flow for the disconnected WPCOM account.
- Rename \<ReconnectAccounts\> and its related functions and components to \<ReconnectGoogleAccount\> accordingly to distinguish them from the new components.
- Forward rest props of \<AccountCard\> to \<Section.Card\>, and adjust its style.

❗ Please note that this PR doesn't include the implementation of disabled menu links.

![image](https://user-images.githubusercontent.com/17420811/161011076-0a7f5552-899b-46a3-b2cb-385c0b2bb058.png)

### Screenshots:

#### 📹  Re-connect WPCOM account

https://user-images.githubusercontent.com/17420811/161008750-173be14c-465c-46fa-9503-bb18fddfd8d2.mp4

#### 📹  Re-connect Google account

https://user-images.githubusercontent.com/17420811/161005331-5edde506-04e0-45f9-a226-d2e46ba65e41.mp4

### Detailed test instructions:

💡  If you are using a local site for testing, please set the site to be accessible from the public network first.

#### Re-connect WPCOM account

1. Go to the Connection Test page `/wp-admin/admin.php?page=connection-test-admin-page`.
2. Click on the "Disconnect Jetpack" button.
3. Go to a GLA page. It should detect the disconnected WPCOM account and redirect to the reconnecting page of WPCOM account.
4. Check if the reconnecting page of WPCOM account is the same as the visual design.
5. Click on the "Connect" and finish the authorization process.
6. After back from the authorization, the page should be at the Settings page and show the connected accounts.

#### Re-connect Google account

1. Go to the Connection Test page `/wp-admin/admin.php?page=connection-test-admin-page`.
2. Click on the "Disconnect Google Account" button.
3. Go to a GLA page. It should detect the disconnected Google account and redirect to the reconnecting page of Google account.
5. Click on the "Connect" and finish the authorization process.
6. After back from the authorization, the page should be at the Dashboard page.

### Additional details:

Please refer to #1283 for the error response when WPCOM (Jetpack) or Google account is required to be reconnected.

Figma link: eQ9O4m2flzrcAiDMXkOW0m-fi-6257%3A116720

### Changelog entry

> Add - Enables the re-connecting flow for the disconnected WordPress.com (Jetpack) account.

eason9487 committed:
Rename reconnect-accounts to reconnect-google-account

eason9487 committed:
Change the URL subpath of reconnecting google account

eason9487 committed:
Add the `next` parameter to GET `jetpack/connect` API for indicating the redirect URI when back from Jetpack authorization

eason9487 committed:
Forward rest props of AccountCard to Section.Card, and adjust its style

eason9487 committed:
Make ConnectWPComAccountCard call `jetpack/connect` API with the `next` parameter.

eason9487 committed:
Add a new component for reconnecting WPCOM account

eason9487 committed:
Add the re-connecting flow for the disconnected WPCOM account

eason9487 commented on js/src/wcdl/section/card/index.js:
```diff
@@ -9,8 +9,8 @@ import { Card as WPCard } from '@wordpress/components';
 
-const Card = ( props ) => {
-	return <WPCard { ...props } size="none" />;
+const Card = ( { size = '', ...props } ) => {
```
📝 
`'none'` is not a valid size value, so changed it to an empty string to avoid appearing `is-size-none` CSS class in the card nodes.

![2022-03-31 16 42 45](https://user-images.githubusercontent.com/17420811/161015624-c4c36e82-7e43-414b-9943-2e4f56c6898e.png)

eason9487 commented:
The shared branch does not yet contain the #1352 fix, so the PHP Unit Tests job on GitHub Actions failed. I was able to run it locally with PHP Unit Test. I was able to run PHP unit tests locally and it passed.

puntope commented on js/src/components/wpcom-account-card/connect-wpcom-account-card.js:
```diff
@@ -2,19 +2,25 @@
 const ConnectWPComAccountCard = () => {
 	const { createNotice } = useDispatchCoreNotices();
+
+	const next = glaData.mcSetupComplete ? 'reconnect' : 'setup-mc';
```
💅 

Wondering about the naming of "next" which sounds a bit generic to me. But I guess it comes from backend implementation

eason9487 replied:
Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/1385/commits/cabe35a5033fec920e3d91bcd96c011569f9cc1a.

puntope commented on js/src/data/index.js:
```diff
@@ -40,6 +41,18 @@ apiFetch.use(
 				.then( ( errorInfo ) => {
+					const { code } = errorInfo;
+
+					if ( code === 'JETPACK_DISCONNECTED' ) {
+						getHistory().replace( getReconnectWPComAccountUrl() );
+					} else if ( code === 'GOOGLE_DISCONNECTED' ) {
+						getHistory().replace( getReconnectGoogleAccountUrl() );
+					}
```
💅 WDYT of saving `JETPACK_DISCONNECTED` and `GOOGLE_DISSCONECTED`as constants?

eason9487 replied:
Added in https://github.com/woocommerce/google-listings-and-ads/pull/1385/commits/9eb735608b1945a2d546d9421cdcf1d426211923.

puntope commented on js/src/data/index.js:
```diff
@@ -40,6 +41,18 @@ apiFetch.use(
+						getHistory().replace( getReconnectWPComAccountUrl() );
+					} else if ( code === 'GOOGLE_DISCONNECTED' ) {
+						getHistory().replace( getReconnectGoogleAccountUrl() );
+					}
```
💅 WDYT about having

`getReconnectAccountUrl( code )`

instead of the conditional approach

?

eason9487 replied:
Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/1385/commits/9eb735608b1945a2d546d9421cdcf1d426211923 with another suggestion. Eventually, it still needs a `if` checking here for not to handle other error codes.

puntope commented on js/src/utils/urls.js:
```diff
@@ -58,9 +59,17 @@ export const getEditStoreAddressUrl = () => {
 };
 
-export const getReconnectAccountsUrl = () => {
+export const getReconnectWPComAccountUrl = () => {
 	return getNewPath(
-		{ subpath: subpaths.reconnectAccounts },
+		{ subpath: subpaths.reconnectWPComAccount },
+		settingsPath,
+		null
+	);
+};
+
+export const getReconnectGoogleAccountUrl = () => {
+	return getNewPath(
+		{ subpath: subpaths.reconnectGoogleAccount },
 		settingsPath,
 		null
 	);
```
💅 

Related to https://github.com/woocommerce/google-listings-and-ads/pull/1385/files#r839576276

WDYT of 

```
export const getReconnectAccountUrl = ( type ) => {
let subpath = subpaths.reconnectWPComAccount;

if (type === GOOGLE_DISCONNECTED) {
   subpath = subpaths.reconnectGoogleAccount;
}

return getNewPath(
		{ subpath },
		settingsPath,
		null
	);

}

```

eason9487 replied:
Adjusted in https://github.com/woocommerce/google-listings-and-ads/commit/9eb735608b1945a2d546d9421cdcf1d426211923.

puntope commented on src/Tracking/README.md:
```diff
@@ -269,6 +269,7 @@ All event names are prefixed by `wcadmin_gla_`.
     -   `id`: (tooltip identifier)
 
 -   `wordpress_account_connect_button_click` - Clicking on the button to connect WordPress.com account.
+    -   `context`: (`setup-mc`|`reconnect`) - indicate the button is clicked from which page.
```
```suggestion
    -   `context`: (`setup-mc`|`reconnect`) - indicates from which page the button was clicked.
```

puntope commented on src/API/Site/Controllers/Jetpack/AccountController.php:
```diff
@@ -114,6 +130,24 @@ protected function get_connect_callback(): callable {
+		return [
+			'context' => $this->get_context_param( [ 'default' => 'view' ] ),
+			'next'    => [
+				'description'       => __( 'Indicate the next page name to map the redirect URI when back from Jetpack authorization.', 'google-listings-and-ads' ),
```
Indicates* (3rd singular person)

puntope replied:
Suggestion-> Indicates the URL Callback after Jetpack authorization

eason9487 replied:
Adjusted in https://github.com/woocommerce/google-listings-and-ads/pull/1385/commits/cabe35a5033fec920e3d91bcd96c011569f9cc1a. The value is a page name but not a URL, so I kept "the next page name" in the description. Hope it's clearer now.

puntope commented on src/API/Site/Controllers/Jetpack/AccountController.php:
```diff
@@ -30,6 +31,18 @@ class AccountController extends BaseOptionsController {
+	 *
+	 * @var string[]
+	 */
+	private const NEXT_PATH_MAPPING = [
```
💅  About `next` naming in general 

Was not clear at first glance what that `next` param was when I looked at the frontend. I needed to read the docs and the code to finally understood that was meant the URL Callback provided by the Authentication Jetpack API.

WDYT about `url_callback`? Seems like its how they named those in [Auth0](https://auth0.github.io/auth0-oidc-client-net/documentation/getting-started/callbacks.html)

eason9487 replied:
Same as https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r840314974.

puntope reviewed (commented):
✅  Behaviour works as expected, thanks :) 

-> 💅  Some comments regarding code to discuss.

eason9487 committed:
Update a description in the tracking README.md

Address https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r839588406

Co-authored-by: Miguel Pérez Pellicer <5908855+puntope@users.noreply.github.com>

eason9487 committed:
Rename the `next` parameters of Google and JetPack connect APIs to `next_page_name`.

Address
- https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r839567333
- https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r839589560
- https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r839596788

eason9487 committed:
Combine two functions that get the re-connecting account URL into one.

Address
- https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r839569258
- https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r839576276
- https://github.com/woocommerce/google-listings-and-ads/pull/1385#discussion_r839580484

eason9487 committed:
Fix PHP linting warnings

puntope reviewed (approved):
✅  Good to go (still not sure about that "nextPage" naming but It's not a blocker)
